trie/trie.go

```diff
@@ -464,74 +460,119 @@ func newHasher() *hasher {
-// size of the child is larger than a hash.
-func (h *hasher) replaceChildren(n node, db DatabaseWriter) (node, error) {
+// hashChildren replaces the children of a node with their hashes if the encoded
+// size of the child is larger than a hash, returning the collapsed node as well
+// as a replacement for the original node with the child hashes cached in.
+func (h *hasher) hashChildren(original node, db DatabaseWriter) (node, node, error) {
 	var err error
-	switch n := n.(type) {
+
+	switch n := original.(type) {
 	case shortNode:
+		// Hash the short node's child, caching the newly hashed subtree
+		cached := shortNode{Key: common.CopyBytes(n.Key), Val: n.Val, dirty: true}
+		if n.hash != nil {
+			cached.dirty = n.dirty
+		}
 		n.Key = compactEncode(n.Key)
 		if _, ok := n.Val.(valueNode); !ok {
-			if n.Val, err = h.hash(n.Val, db, false); err != nil {
-				return n, err
+			if n.Val, cached.Val, err = h.hash(n.Val, db, false); err != nil {
+				return n, original, err
 			}
 		}
 		if n.Val == nil {
-			// Ensure that nil children are encoded as empty strings.
-			n.Val = valueNode(nil)
+			n.Val = valueNode(nil) // Ensure that nil children are encoded as empty strings.
 		}
-		return n, nil
+		return n, cached, nil
+
 	case fullNode:
+		// Hash the full node's children, caching the newly hashed subtrees
+		cached := fullNode{dirty: true}
+		if n.hash != nil {
+			cached.dirty = n.dirty
+		}
 		for i := 0; i < 16; i++ {
-			if n[i] != nil {
-				if n[i], err = h.hash(n[i], db, false); err != nil {
-					return n, err
+			if n.Children[i] != nil {
+				if n.Children[i], cached.Children[i], err = h.hash(n.Children[i], db, false); err != nil {
+					return n, original, err
 				}
 			} else {
-				// Ensure that nil children are encoded as empty strings.
-				n[i] = valueNode(nil)
+				n.Children[i] = valueNode(nil) // Ensure that nil children are encoded as empty strings.
 			}
 		}
-		if n[16] == nil {
-			n[16] = valueNode(nil)
+		cached.Children[16] = n.Children[16]
+		if n.Children[16] == nil {
+			n.Children[16] = valueNode(nil)
 		}
-		return n, nil
+		return n, cached, nil
+
 	default:
-		return n, nil
+		// Value and hash nodes don't have children so they're left as were
+		return n, original, nil
 	}
 }
 
-func (h *hasher) store(n node, db DatabaseWriter, force bool) (node, error) {
-	// Don't store hashes or empty nodes.
+func (h *hasher) store(n node, db DatabaseWriter, force bool) (node, []byte, error) {
+	// Don't store hashes or empty nodes
 	if _, isHash := n.(hashNode); n == nil || isHash {
-		return n, nil
+		return n, nil, nil
 	}
-	h.tmp.Reset()
-	if err := rlp.Encode(h.tmp, n); err != nil {
-		panic("encode error: " + err.Error())
+	// Retrieve any cached values and only recompute missing ones
+	hash, blob := n.cachedData()
+
+	if blob == nil {
+		h.tmp.Reset()
+		if err := rlp.Encode(h.tmp, n); err != nil {
+			panic("encode error: " + err.Error())
+		}
+		blob = common.CopyBytes(h.tmp.Bytes())
```

Comment:
I don't see the benefit of caching the RLP encoding here. It's a memory/cpu tradeoff. For most non-leaf nodes caching the RLP encoding is useless because it will be invalidated with the next change. Even for nodes which are not modified it increases the amount of memory that is wasted.

I added the tmp buffer reuse to save memory allocations. If you want to have a fresh []byte for each node you might as well use `rlp.EncodeToBytes` instead of using tmp and then copying tmp.Bytes. The idea is that the database operation will copy the buffer so it is not needed beyond the db.Put call.